File: dtk_report.pl
#!/usr/bin/perl
use warnings;
use strict;
use File::Find qw/find/;
use autodie qw/open close/;
use Cwd qw/abs_path/;

if (scalar(@ARGV) == 0) {
  print "run ./dtk -d", $/;
	exit 1;
}

## verify the timer:
## the time must be between 0 and 23
if ( $ARGV[0] < 0 || $ARGV[0] > 23 ) {
    print "Enter a valid report time between 0 & 23.\n";
    exit 1;
}

## search to see if incoming and outgoing
## trunk group files exist

## files to seek
my $files = [];

# search and get the name of files
find(
    sub {
        push @$files => abs_path($_) if /\.*?\.csv$/i;
    },
    "datafiles"
);

if ( scalar(@$files) != 2 ) {
    print "There must be atleast two files in 'datafiles' directory.\n";
    exit 1;
}
else {
    my $counter = 0;
    for (@$files) {
        $counter++
          if
/ne_5_Incoming_Calls_through_Trunk_Groups|ne_5_Outgoing_Calls_through_Trunk_Groups/;
    }
    exit 1 if $counter != 2;
}

collate( read_file($files) );

sub read_file {
    my $files = shift;

    my $content = {};
    my $header;

    for my $file ( @{$files} ) {
        open my $fh, '<', "$file";
        while (<$fh>) {
            if (/.?((outgo|incom)ing)/i) {
                $header = $1;
            }
            else {
                my $data = [ split /,/ ];

                ## modified the date
                my $day =
                  @{ [ split /\s+?/ => @$data[2] ] }[0];
                $day =~ s/"(.+)/$1/;

                ## modified name
                ( my $name = @$data[0] ) =~ s/"(.*)\(.*$/$1/img;

                ## get content for each hour
                # $ARGV[0] is the timer specified on the CLI
                push @{ $content->{$day}{$header}{$name} } =>
                  [ @$data[ 4 .. 9 ] ]
                  if $data->[2] =~ /$ARGV[0]:00:00/;
            }
        }
        close $fh;
    }
    return $content;
}

sub collate {

    # report file
    my $filename = 'Trunk Report_';
    open my $fh, '>', "./${filename}$$.xls";
    my $data = shift;
    for my $day ( sort { $a cmp $b } keys %{$data} ) {
        print $fh $day, $/;
        for my $key ( sort { $a cmp $b } keys %{ $data->{$day} } ) {
            print $fh $key, $/;
            print $fh title(), $/;
            for my $name ( sort { $a cmp $b } keys %{ $data->{$day}{$key} } ) {
                print $fh $name, "\t",
                  join( "\t" => @{ @{ $data->{$day}{$key}{$name} }[0] } ), $/;
            }
        }
    }
    close $fh;
}

sub title {
    my $msg =
"Incoming or bidirection trunk group|Number of Installed Circuits (PIECES)|";
    $msg .=
"Number of Available Circuits (PIECES)|Number of Blocked Circuits (PIECES)|";
    $msg .=
"Call Completion Rate (PERCENT)|Answer Rate (PERCENT)	Seizure Traffic (ERL";
    return join( "\t" => @{ [ split /\|/ => $msg ] } );
}

File: dtk_report.py
#!/usr/bin/env python3

import os
import sys
import unittest

__AUTHOR__ = 'emrys'
__VERSION__ = 0.1


class NElement:
    hourly_content: list[str] = []

    def __init__(self, name: str) -> None:
        self._name = name

    def __repr__(self) -> str:
        return '|'.join(self.hourly_content)

    def __str__(self) -> str:
        return self.__repr__()

    def _get_data(self, line: str) -> None:
        self.hourly_content.append(line)


def read_file(files: list[str]) -> None:
    import re
    os.chdir('./datafiles')
    # print('TODO: Implement container and parse each line')
    for file in files:
        with open(file) as fin:
            for line in fin.readlines():
                for wrd in line.split(','):
                    if (re.match(r'^("Incom|"Outgo)ing', wrd)):
                        print(wrd.split(' ')[0].replace('"', ''))
                    elif (re.search(r'\(\d+?\)', wrd, re.DOTALL)):
                        wrd = re.search(r'"(.*)\(', wrd).groups(1)[0]
                        print(wrd)
                    # else:
                    #     print(wrd)


def search_locate(dir: str) -> tuple[int, list[str]]:
    'search to find if csv files exist in the specified location.'
    # check if the directory exist
    if not os.path.exists(dir):
        print(f'Path specified - {dir} doesn\'t exist.')
        sys.exit(1)

    files = \
        list(file
             for file in os.listdir(dir) if file.endswith('.csv'))
    return (len(files), files)


def main(lst: list[str]) -> None:
    'Entry point for all function.'
    try:
        hour = int(lst[0])
        if (hour < 0 or hour > 23):
            print('Enter a valid report time between 0 & 23')
            sys.exit(1)
    except Exception as ex:
        print(f'Usage: {sys.argv[0]} [timer]')
        sys.exit(1)

    number_file, list_file = search_locate('datafiles')
    if (number_file) != 2:
        file_names = \
            f'/ne_5_Incoming_Calls_through_Trunk_Groups|ne_5_Outgoing_Calls_through_Trunk_Groups/'
        print(file_names+' must in the directory `datafiles`')
        sys.exit(1)
    else:
        read_file(list_file)


if __name__ == '__main__':
    main(sys.argv[1:])


class TestingClass(unittest.TestCase):
    def test_repr_fn(self) -> None:
        ne = NElement('')
        for num in list(range(0, 10, 2)):
            ne._get_data(str(num))
        self.assertEqual(ne.__repr__(), [1, 2, 3, 4, 5])
